### PR TITLE
Extract eval code from GPTQ for more general usage

### PR DESCRIPTION
**Pull request by andrewor14**
Summary: This commit extracts all the eval code from GPTQ.py. This is the first step towards having a general eval framework in torchao. The eventual goal is to use lm_eval to produce reproducible benchmarks for the quantization APIs in torchao that we can showcase on the main README. This will have the added benefit of allowing us to add (possibly nightly) regression test suites for important models.

Test Plan:

```
2024-05-24:14:50:32,647 INFO     [task.py:395] Building contexts for wikitext on rank 0...
100%|██████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 1/1 [00:00<00:00, 1132.98it/s]
2024-05-24:14:50:32,648 INFO     [evaluator.py:362] Running loglikelihood_rolling requests
100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 1/1 [00:51<00:00, 51.39s/it]
wikitext: {'word_perplexity,none': 7.877762491958485, 'word_perplexity_stderr,none': 'N/A', 'byte_perplexity,none': 1.488984329919892, 'byte_perplexity_stderr,none': 'N/A', 'bits_per_byte,none': 0.5743285710685551, 'bits_per_byte_stderr,none': 'N/A', 'alias': 'wikitext'}
.
----------------------------------------------------------------------
Ran 1 test in 858.105s

OK
```

python test/quantization/test_quant_api.py -k test_8da4w_gptq_quantizer

### CODE REVIEW

File: test/quantization/test_quant_api.py [hunk 1]
```diff
@@ -186,9 +186,14 @@ def test_8da4w_quantizer(self):
         assert isinstance(m.linear2, Int8DynActInt4WeightLinear)
         m(*example_inputs)
 
+    # TODO: save model weights as artifacts and re-enable in CI
```
Comment (msaroufim) [inline test/quantization/test_quant_api.py:189]: Until this is running in CI mind if we have an `eval.py` we can run in a new `scripts/` folder?

File: torchao/quantization/_eval.py [hunk 2]
```diff
@@ -0,0 +1,228 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates.
```
Comment (msaroufim) [inline torchao/quantization/_eval.py:1]: Can we put this in a higher level namespace? Something like torchao._eval? or torchao.model._eval or torchao.util

File: torchao/quantization/GPTQ.py [hunk 3]
```diff
@@ -32,266 +37,20 @@
...
     add_ons += ["Int8DynActInt4WeightQuantizer", "Int8DynActInt4WeightGPTQQuantizer"]
 
 
 __all__ = [
-    "MultiInput",
     "Int4WeightOnlyGPTQQuantizer",
     "Int4WeightOnlyQuantizer",
 ] + add_ons
 
-if lm_eval_available:
```
Comment (msaroufim) [inline torchao/quantization/GPTQ.py:73]: awesome <3 thank you for making this change

**Review by msaroufim — APPROVED** [review]

File: test/quantization/test_quant_api.py (hunk 1 above)
Comment (andrewor14) [reply test/quantization/test_quant_api.py:189]: Ok, will do in the next PR. Currently the models still live under test/ so we'll probably have to move those out as well

File: torchao/quantization/utils.py [hunk 4]
```diff
@@ -22,6 +22,13 @@
     "TORCH_VERSION_AFTER_2_3",
 ]
 
+try:
+    import lm_eval  # pyre-ignore[21]  # noqa: F401
+
+    _lm_eval_available = True
+except:
+    _lm_eval_available = False
+
```
Comment (sanchitintel) [inline torchao/quantization/utils.py:31]: With python 3.8, importing `_lm_eval_available` is failing. Could we add `_lm_eval_available = False` before the `try-except` block? Thanks!

```
ImportError: cannot import name '_lm_eval_available' from 'torchao.utils'
```

File: torchao/_eval.py [hunk 5]
```diff
@@ -0,0 +1,228 @@
...
+# This source code is licensed under the BSD-style license found in the
+# LICENSE file in the root directory of this source tree.
+
+
+# This source code is licensed under the license found in the
+# LICENSE file in the root directory of this source tree.
+
+import torch
+
+from .utils import _lm_eval_available, _MultiInput
```
Comment (sanchitintel) [inline torchao/_eval.py:13]: These two are in a different directory.

Comment (andrewor14) [reply torchao/_eval.py:13]: thanks for catching that, will submit a fix

**Pull request merged**